providers/implementations/kem/ml_kem.c

```diff
@@ -21,212 +21,236 @@
     }
 
-    if (ctx->key == NULL
-        || ctx->key->keytype != MLKEM_KEY_TYPE_768
-        || ctx->key->encoded_pubkey == NULL
-        || secret == NULL)
-        return 0;
+    if (*secretlen < MLKEM_SHARED_SECRET_BYTES) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
+                       "short ML-KEM encapsulate shared secret");
+        goto end;
+    }
+    if (*outlen < v->ctext_bytes) {
+        ERR_raise_data(ERR_LIB_PROV, PROV_R_BAD_LENGTH,
+                       "short ML-KEM encapsulate ciphertext");
+        goto end;
+    }
 
-    if (ctx->entropy != NULL) {
-        ret = ossl_mlkem768_encap_external_entropy(out, secret,
-                                                   &ctx->key->pubkey,
-                                                   ctx->entropy,
-                                                   ctx->key->mlkem_ctx);
-    } else {
-        ret = ossl_mlkem768_encap(out, (uint8_t *)secret, &ctx->key->pubkey,
-                                  ctx->key->mlkem_ctx);
+    if (secret == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_SECRET);
+        goto end;
     }
 
-    debug_print("MLKEMKEM encaps returns %d\n", ret);
+    *secretlen = MLKEM_SHARED_SECRET_BYTES;
+    *outlen = key->vinfo->ctext_bytes;
+    if (ctx->have_entropy)
+        ret = ossl_mlkem_vencap_seed(v, out, *outlen, secret, *secretlen,
+                                     key->pubkey, key->prvkey, ctx->entropy,
+                                     sizeof(ctx->entropy), key->ctx);
+    else
+        ret = ossl_mlkem_vencap_rand(v, out, *outlen, secret, *secretlen,
+                                     key->pubkey, key->prvkey, key->ctx);
+
+  end:
+    if (ctx->have_entropy) {
+        OPENSSL_cleanse(ctx->entropy, sizeof(ctx->entropy));
+        ctx->have_entropy = 0;
+    }
     return ret;
 }
 
 static int mlkem_decapsulate(void *vctx, unsigned char *out, size_t *outlen,
                              const unsigned char *in, size_t inlen)
 {
-    PROV_MLKEM_CTX *ctx = (PROV_MLKEM_CTX *)vctx;
-    int ret;
+    PROV_MLKEM_CTX *ctx = vctx;
+    MLKEM_KEY *key = ctx->key;
+    vinfo_t v;
 
-    debug_print("MLKEMKEM decaps %p to %p\n", ctx, out);
-    debug_print("MLKEMKEM decaps inlen at %ld\n", inlen);
-    if (outlen != NULL)
-        *outlen = OSSL_MLKEM768_SHARED_SECRET_BYTES;
+    if (key == NULL || key->prvkey == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return 0;
+    }
+    v = key->vinfo;
 
     if (out == NULL) {
-        debug_print("MLKEMKEM decaps outlen set to %ld \n", *outlen);
+        if (outlen == NULL)
+            return 0;
+        *outlen = MLKEM_SHARED_SECRET_BYTES;
         return 1;
     }
 
-    if (ctx->key == NULL
-        || ctx->key->keytype != MLKEM_KEY_TYPE_768
-        || ctx->key->encoded_privkey == NULL
-        || in == NULL)
-        return 0;
-
-    if (inlen != OSSL_MLKEM768_CIPHERTEXT_BYTES)
+    if (*outlen < MLKEM_SHARED_SECRET_BYTES) {
```

Comment:
And what if *outlen > MLKEM_SHARED_SECRET_BYTES?

Reply:
That's OK, we return the actual size. It is expected that providers tolerate larger sizes than requested.